src/envoy/http/authn/origin_authenticator_test.cc

```diff
@@ -0,0 +1,256 @@
+using testing::SetArgPointee;
+using testing::StrictMock;
+
+namespace Envoy {
+namespace Http {
+namespace Istio{
+namespace AuthN {
+namespace {
+
+const char kSingleMethodRule[] = R"(
+  binding: USE_ORIGIN
+  origins {
+    jwt {
+      issuer: "abc.xyz"
+    }
+  }
+)";
+
+const char kMultipleMethodsRule[] = R"(
+  binding: USE_ORIGIN
+  origins {
+    jwt {
+      issuer: "one"
+    }
+  }
+  origins {
+    jwt {
+      issuer: "two"
+    }
+  }
+  origins {
+    jwt {
+      issuer: "three"
+    }
+  }
+)";
+
+const char kPeerBinding[] = R"(
+  binding: USE_PEER
+  origins {
+    jwt {
+      issuer: "abc.xyz"
+    }
+  }
+)";
+
+class MockAuthenticator : public OriginAuthenticator {
+ public:
+  MockAuthenticator(FilterContext* filter_context,
+                    const DoneCallback& done_callback,
+                    const iaapi::CredentialRule& rule)
+      : OriginAuthenticator(filter_context, done_callback, rule) {}
+
+  MOCK_CONST_METHOD2(validateX509,
+                     void(const iaapi::MutualTls&, const MethodDoneCallback&));
+  MOCK_CONST_METHOD2(validateJwt,
+                     void(const iaapi::Jwt&, const MethodDoneCallback&));
+};
+
+class OriginAuthenticatorTest : public testing::TestWithParam<bool> {
+ public:
+  OriginAuthenticatorTest()
+      : request_headers_{{":method", "GET"}, {":path", "/"}} {}
+  virtual ~OriginAuthenticatorTest() {}
+
+  void SetUp() override {
+    filter_context_.setHeaders(&request_headers_);
+    expected_result_when_pass_ = TestUtilities::AuthNResultFromString(R"(
+      principal: "foo"
+      origin {
+        user: "foo"
+        presenter: "istio.io"
+      }
+    )");
+    set_peer_ = GetParam();
+    if (set_peer_) {
+      auto peer_result = TestUtilities::CreateX509Payload("bar");
+      filter_context_.setPeerResult(&peer_result);
+      expected_result_when_pass_.set_peer_user("bar");
+    }
+    initial_result_ = filter_context_.authenticationResult();
+  }
+
+  void init() {
+    authenticator_.reset(new StrictMock<MockAuthenticator>(
+        &filter_context_, on_done_callback_.AsStdFunction(), rule_));
+  }
+
+ protected:
+  std::unique_ptr<StrictMock<MockAuthenticator>> authenticator_;
+  StrictMock<TestUtilities::MockFilterContext> filter_context_;
+  StrictMock<MockFunction<void(bool)>> on_done_callback_;
+  Http::TestHeaderMapImpl request_headers_;
+  iaapi::CredentialRule rule_;
+
+  // Mock response payload.
+  Payload jwt_payload_{TestUtilities::CreateJwtPayload("foo", "istio.io")};
+  // Expected result (when authentication pass with mock payload above)
+  Result expected_result_when_pass_;
+  // Copy of authN result (from filter context) before running authentication.
+  // This should be the expected result if authn fail or do nothing.
+  Result initial_result_;
+
+  // Indicates peer is set in the authN result before running. This is set from
+  // test GetParam()
+  bool set_peer_;
+};
+
+TEST_P(OriginAuthenticatorTest, Empty) {
+  init();
```

Comment:
why there is Setup() and init()?   Shouldn't they be combined?

Reply:
The rule / policy is modified for each test, so authenticator_ setup need to be called outside Setup.

Reply:
then, rename init() to be more specific

Reply:
done